daemon/api_quota.go

```diff
@@ -0,0 +1,163 @@
+		Path: "/v2/quotas/{group}",
+		GET:  getQuotaGroupInfo,
+	}
+)
+
+type postQuotaGroupData struct {
+	// Action can be "ensure" or "remove"
+	Action    string   `json:"action"`
+	GroupName string   `json:"group-name"`
+	MaxMemory uint64   `json:"max-memory,omitempty"`
+	Parent    string   `json:"parent,omitempty"`
+	Snaps     []string `json:"snaps,omitempty"`
+}
+
+type quotaGroupResultJSON struct {
+	GroupName string   `json:"group-name"`
+	MaxMemory uint64   `json:"max-memory"`
+	Parent    string   `json:"parent,omitempty"`
+	Snaps     []string `json:"snaps,omitempty"`
+	SubGroups []string `json:"subgroups,omitempty"`
+}
+
+var (
+	servicestateCreateQuota = servicestate.CreateQuota
+	servicestateRemoveQuota = servicestate.RemoveQuota
+)
+
+// getQuotaGroups returns all quota groups sorted by name.
+func getQuotaGroups(c *Command, r *http.Request, _ *auth.UserState) Response {
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	quotas, err := servicestate.AllQuotas(st)
+	if err != nil {
+		return InternalError("cannot get quotas: %v", err)
+	}
+
+	i := 0
+	names := make([]string, len(quotas))
+	for name := range quotas {
+		names[i] = name
+		i++
+	}
+	sort.Strings(names)
+
+	results := make([]quotaGroupResultJSON, len(quotas))
+	for i, name := range names {
+		qt := quotas[name]
+		results[i] = quotaGroupResultJSON{
+			GroupName: qt.Name,
+			Parent:    qt.ParentGroup,
+			SubGroups: qt.SubGroups,
+			Snaps:     qt.Snaps,
+			MaxMemory: uint64(qt.MemoryLimit),
+		}
+	}
+	return SyncResponse(results, nil)
+}
+
+// getQuotaGroupInfo returns details of a single quota Group.
+func getQuotaGroupInfo(c *Command, r *http.Request, _ *auth.UserState) Response {
+	vars := muxVars(r)
+	groupName := vars["group"]
+	if err := naming.ValidateQuotaGroup(groupName); err != nil {
+		return BadRequest(err.Error())
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	quotas, err := servicestate.AllQuotas(st)
+	if err != nil {
+		return InternalError("cannot get quotas: %v", err)
+	}
+	group := quotas[groupName]
+	if group == nil {
+		return NotFound("cannot find quota group %q", groupName)
+	}
+
+	res := quotaGroupResultJSON{
+		GroupName: group.Name,
+		Parent:    group.ParentGroup,
+		Snaps:     group.Snaps,
+		SubGroups: group.SubGroups,
+		MaxMemory: uint64(group.MemoryLimit),
+	}
+	return SyncResponse(res, nil)
+}
+
+// postQuotaGroup creates quota resource group or updates an existing group.
+func postQuotaGroup(c *Command, r *http.Request, _ *auth.UserState) Response {
+	var data postQuotaGroupData
+
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&data); err != nil {
+		return BadRequest("cannot decode create-user data from request body: %v", err)
+	}
+
+	if err := naming.ValidateQuotaGroup(data.GroupName); err != nil {
+		return BadRequest(err.Error())
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	switch data.Action {
+	case "ensure":
+		// TODO: quota updates
```

Comment:
Given that this will error is trying to manipulate an existing group I think it's fine for now. Doing the manipulation in a followup is fine. This will allow us to basic the minimal feature working end-to-end.